[pytorch] add '__BASE__' section to op deps to factor out frequently used util ops

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37404 [pytorch] add '__BASE__' section to op deps to factor out frequently used util ops**
* #37393 [pytorch] remove unused flags from code analyzer & move format support to python

Summary:
Many aten operators are really like util functions, e.g.:
aten::is_nonzero, aten::is_floating_point, etc. These ops can be called
via overloaded c++ operator, so seemingly trivial and innocent code changes can
affect how these ops are used by other ops (thus changes the output of
static analyzer).

Most of these util ops are rather small in terms of build size cost, so
for the purpose of optimizing binary size with custom build, whether to
include these ops or not does not make significant difference. In fact
for non-trivial models a set of these ops are almost always used.

This PR introduced the (optional) '__BASE__' ops section to the dependency graph.

We can maintain the list of frequently used small util ops for internal BUCK
build. This way, the output dependency graph will only contain meaningful
edges with significant binary size impact, and it will be more stable from
trivial code changes (which is checked in FB codebase).

Having a stable and sparse deps graph by factoring out frequently used based ops
is also a nice property to allow us to explore alternative custom build
solutions in case we find it hard to maintain the static code analyzer.

Differential Revision: [D21280835](https://our.internmc.facebook.com/intern/diff/D21280835)

ljk53 committed:
[pytorch] add '__BASE__' section to op deps to factor out frequently used util ops

Summary:
Many aten operators are really like util functions, e.g.:
aten::is_nonzero, aten::is_floating_point, etc. These ops can be called
via overloaded c++ operator, so seemingly trivial and innocent code changes can
affect how these ops are used by other ops (thus changes the output of
static analyzer).

Most of these util ops are rather small in terms of build size cost, so
for the purpose of optimizing binary size with custom build, whether to
include these ops or not does not make significant difference. In fact
for non-trivial models a set of these ops are almost always used.

This PR introduced the (optional) '__BASE__' ops section to the dependency graph.

We can maintain the list of frequently used small util ops for internal BUCK
build. This way, the output dependency graph will only contain meaningful
edges with significant binary size impact, and it will be more stable from
trivial code changes (which is checked in FB codebase).

Having a stable and sparse deps graph by factoring out frequently used based ops
is also a nice property to allow us to explore alternative custom build
solutions in case we find it hard to maintain the static code analyzer.

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add '__BASE__' section to op deps to factor out frequently used util ops"


Summary:
Many aten operators are really like util functions, e.g.:
aten::is_nonzero, aten::is_floating_point, etc. These ops can be called
via overloaded c++ operator, so seemingly trivial and innocent code changes can
affect how these ops are used by other ops (thus changes the output of
static analyzer).

Most of these util ops are rather small in terms of build size cost, so
for the purpose of optimizing binary size with custom build, whether to
include these ops or not does not make significant difference. In fact
for non-trivial models a set of these ops are almost always used.

This PR introduced the (optional) '__BASE__' ops section to the dependency graph.

We can maintain the list of frequently used small util ops for internal BUCK
build. This way, the output dependency graph will only contain meaningful
edges with significant binary size impact, and it will be more stable from
trivial code changes (which is checked in FB codebase).

Having a stable and sparse deps graph by factoring out frequently used based ops
is also a nice property to allow us to explore alternative custom build
solutions in case we find it hard to maintain the static code analyzer.

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add '__BASE__' section to op deps to factor out frequently used util ops"


Summary:
Many aten operators are really like util functions, e.g.:
aten::is_nonzero, aten::is_floating_point, etc. These ops can be called
via overloaded c++ operator, so seemingly trivial and innocent code changes can
affect how these ops are used by other ops (thus changes the output of
static analyzer).

Most of these util ops are rather small in terms of build size cost, so
for the purpose of optimizing binary size with custom build, whether to
include these ops or not does not make significant difference. In fact
for non-trivial models a set of these ops are almost always used.

This PR introduced the (optional) '__BASE__' ops section to the dependency graph.

We can maintain the list of frequently used small util ops for internal BUCK
build. This way, the output dependency graph will only contain meaningful
edges with significant binary size impact, and it will be more stable from
trivial code changes (which is checked in FB codebase).

Having a stable and sparse deps graph by factoring out frequently used based ops
is also a nice property to allow us to explore alternative custom build
solutions in case we find it hard to maintain the static code analyzer.

Differential Revision: [D21280835](https://our.internmc.facebook.com/intern/diff/D21280835)

[ghstack-poisoned]

bhosmer commented on tools/code_analyzer/gen_op_registration_whitelist.py:
```diff
@@ -42,6 +42,10 @@ def gen_transitive_closure(dep_graph, root_ops):
 
+    # The dependency graph might contain a special entry with key = `__BASE__`
+    # and value = (set of `base` ops to always include in custom build).
+    queue.append('__BASE__')
```
Should we insist that the set of base ops is closed? IIUC adding them here lets us handle a base op generating a non-base dependency, but do we need/want to do that?

ljk53 replied:
@bhosmer maybe, having an explicit closure can prevent us from inadvertently include more ops than what we expect; but on the other hand I feel we will need take extra step to verify the base ops is a closure (to avoid more subtle runtime error). I lean towards the current approach as including too many ops can be captured by build size bot.

bhosmer replied:
Yeah, it's true that verifying would be an explicit step - basically a preliminary run through the loop that follows. (And definitely agreed that letting a missed dependency cause a runtime error is suboptimal :P) Sounds like the build size bot is the safety net, so maintaining a closure invariant is probably overkill.

bhosmer commented on tools/code_analyzer/build.sh:
```diff
@@ -107,9 +107,18 @@ convert_output_to_bazel() {
+
+  if [ -n "${BASE_OPS_FILE}" ] && [ -f "${BASE_OPS_FILE}" ]; then
+    args+=(
+      --base_ops $(< ${BASE_OPS_FILE})
```
This is just a list of op names, right? Is it worth giving the format anywhere in this code, or is that documented elsewhere/understood by everybody?

ljk53 replied:
@bhosmer the format of "base_ops" option is documented in the python script; the bash shell simply cat the file content into the command line.
I'm still working on the integration so this is subject to change. In terminal state the python script will likely be called from some other shell script in fbcode.

bhosmer reviewed (approved):
LG!

ljk53 committed:
Update on "[pytorch] add '__BASE__' section to op deps to factor out frequently used util ops"


Summary:
Many aten operators are really like util functions, e.g.:
aten::is_nonzero, aten::is_floating_point, etc. These ops can be called
via overloaded c++ operator, so seemingly trivial and innocent code changes can
affect how these ops are used by other ops (thus changes the output of
static analyzer).

Most of these util ops are rather small in terms of build size cost, so
for the purpose of optimizing binary size with custom build, whether to
include these ops or not does not make significant difference. In fact
for non-trivial models a set of these ops are almost always used.

This PR introduced the (optional) '__BASE__' ops section to the dependency graph.

We can maintain the list of frequently used small util ops for internal BUCK
build. This way, the output dependency graph will only contain meaningful
edges with significant binary size impact, and it will be more stable from
trivial code changes (which is checked in FB codebase).

Having a stable and sparse deps graph by factoring out frequently used based ops
is also a nice property to allow us to explore alternative custom build
solutions in case we find it hard to maintain the static code analyzer.

Differential Revision: [D21280835](https://our.internmc.facebook.com/intern/diff/D21280835)

[ghstack-poisoned]
